modules/phase_field/src/userobjects/PolycrystalVoronoi.C

```diff
@@ -166,6 +209,59 @@ PolycrystalVoronoi::precomputeGrainStructure()
         _centerpoints[grain](2) = _bottom_left(2) + _range(2) * 0.5;
     }
   }
+
+  // Build a KDTree that is used to speedup point search
+  buildSearchTree();
+}
+
+void
+PolycrystalVoronoi::buildSearchTree()
+{
+  if (!_use_kdtree)
+    return;
+
+  auto midplane = _bottom_left + _range / 2.0;
+  dof_id_type lgrianid = 0;
+  _grain_gtl_ids.clear();
+  _grain_gtl_ids.reserve(_centerpoints.size() * std::pow(2.0, _mesh.dimension()));
+  _new_points.clear();
+  // Use more memory. Factor is 2^dim
+  _new_points.reserve(_centerpoints.size() * std::pow(2.0, _mesh.dimension()));
+  // Domain will be extended along the periodic directions
+  // For each direction, a half domain is constructed
+  std::vector<std::vector<Real>> xyzs(3);
+  for (auto & point : _centerpoints)
+  {
+    // Cear up
+    for (unsigned int i = 0; i < 3; i++)
+      xyzs[i].clear();
+
+    // Have at least the original point
+    for (unsigned int i = 0; i < 3; i++)
+      xyzs[i].push_back(point(i));
+
+    // Add new coords when there exists periodic boundary conditions
+    // We extend half domain
+    for (unsigned int i = 0; i < _mesh.dimension(); i++)
+      if (_mesh.isTranslatedPeriodic(_vars[0]->number(), i))
```

Comment:
Should there be a test somewhere for periodic BCs with non-Cartesian displacements, if only to scream and die with a "not supported" message?

Reply:
What function I can use from libmesh to detect non-Cartesian displacements? We do not have any in the MOOSE now. 
It seems our assumption that cartesian displacements are used everywhere.

Reply:
We do non-Cartesian displacements in several reactor type meshes where the periodic displacements are scaled with the radial distance in some fashion. However, I think that's what we are doing with "isTranslatedPeriodic". It should be Ok.

Reply:
It looks to me like isTranslatedPeriodic returns true only if _periodic_dim is set for that variable+dimension, which only happens if addPeriodicVariable is used to add a Cartesian displacement.  It's not querying the PeriodicBoundaries set.